src/optimizer/limit_pushdown.cpp

```diff
@@ -0,0 +1,42 @@
+#include "duckdb/optimizer/limit_pushdown.hpp"
+
+#include "duckdb/planner/operator/logical_limit.hpp"
+#include "duckdb/planner/operator/logical_projection.hpp"
+#include <iostream>
+
+namespace duckdb {
+
+bool LimitPushdown::CanOptimize(duckdb::LogicalOperator &op) {
+	if (op.type == LogicalOperatorType::LOGICAL_LIMIT &&
+	    op.children[0]->type == LogicalOperatorType::LOGICAL_PROJECTION) {
+		auto &limit = op.Cast<LogicalLimit>();
+
+		if (limit.offset_val.Type() == LimitNodeType::EXPRESSION_PERCENTAGE ||
+		    limit.offset_val.Type() == LimitNodeType::EXPRESSION_VALUE) {
+			// Offset cannot be an expression
+			return false;
+		}
+
+		if (limit.limit_val.Type() == LimitNodeType::CONSTANT_VALUE && limit.limit_val.GetConstantValue() < 8192) {
+			// Push down only when limit value is small (< 8192).
+			return true;
+		}
+	}
+	return false;
+}
+
+unique_ptr<LogicalOperator> LimitPushdown::Optimize(unique_ptr<LogicalOperator> op) {
+	if (CanOptimize(*op)) {
+		auto projection = std::move(op->children[0]);
+		op->children[0] = std::move(projection->children[0]);
+		projection->children[0] = std::move(op);
+		swap(projection, op);
```

Comment:
Does the limit get pushed down again? What is you have limit->projection->projection ?

Reply:
Maybe I'm still missing something, but what if the plan looks like
Limit (op) -> Projection -> Projection
After this optimization the plan then looks like

Projection (op) -> Limit -> Projection

But now the optimizer returns. Once if encounters an operator that can be optimized, the children are no longer optimized. 

What I would do is remove the else, and just have 

```
if (CanOptimize(*op)) {
// swap code
}
for (auto &child : op->children) {
	child = Optimize(std::move(child));
}
return op
```

Now if a query is complicated (with many subqueries with limits), this optimization is applied to every limit -> projection pattern.